net/ntske/cookies.go

```diff
@@ -0,0 +1,165 @@
+	return b
+}
+
+func (c *PlainCookie) Decode(b []byte) error {
+	var pos int = 0
+	field_algo, field_s2c, field_c2s := false, false, false
+	for pos < len(b) {
+		var t uint16 = binary.BigEndian.Uint16(b[pos:])
+		var len uint16 = binary.BigEndian.Uint16(b[pos+2:])
+		if t == cookieTypeAlgorithm {
+			c.Algo = binary.BigEndian.Uint16(b[pos+4:])
+			field_algo = true
+		} else if t == cookieTypeKeyS2C {
+			c.S2C = b[pos+4 : pos+4+int(len)]
+			field_s2c = true
+		} else if t == cookieTypeKeyC2S {
+			c.C2S = b[pos+4 : pos+4+int(len)]
+			field_c2s = true
+		}
+		pos += 4 + int(len)
+	}
+	if pos != len(b) {
+		return errors.New("plain cookie has unexpected length")
+	}
+	if !(field_algo && field_s2c && field_c2s) {
+		return errors.New("plain cookie has missing fields")
+	}
+	return nil
+}
+
+type EncryptedCookie struct {
+	ID         uint16
+	Nonce      []byte
+	Ciphertext []byte
+}
+
+func (c *EncryptedCookie) Encode() []byte {
+	var encryptedcookiesize int = 3*4 + 2 + len(c.Nonce) + len(c.Ciphertext)
+	b := make([]byte, encryptedcookiesize)
+	binary.BigEndian.PutUint16(b[0:], cookieTypeKeyID)
+	binary.BigEndian.PutUint16(b[2:], 0x2)
+	binary.BigEndian.PutUint16(b[4:], c.ID)
+	binary.BigEndian.PutUint16(b[6:], cookieTypeNonce)
+	binary.BigEndian.PutUint16(b[8:], uint16(len(c.Nonce)))
+	copy(b[10:], c.Nonce)
+	pos := len(c.Nonce) + 10
+	binary.BigEndian.PutUint16(b[pos:], cookieTypeCiphertext)
+	binary.BigEndian.PutUint16(b[pos+2:], uint16(len(c.Ciphertext)))
+	copy(b[pos+4:], c.Ciphertext)
+	return b
+}
+
+func (c *EncryptedCookie) Decode(b []byte) error {
+	var pos int = 0
+	field_id, field_nonce, field_ciphertext := false, false, false
+	for pos < len(b) {
+		var t uint16 = binary.BigEndian.Uint16(b[pos:])
+		var len uint16 = binary.BigEndian.Uint16(b[pos+2:])
+		if t == cookieTypeKeyID {
+			c.ID = binary.BigEndian.Uint16(b[pos+4:])
+			field_id = true
+		} else if t == cookieTypeNonce {
+			c.Nonce = b[pos+4 : pos+4+int(len)]
+			field_nonce = true
+		} else if t == cookieTypeCiphertext {
+			c.Ciphertext = b[pos+4 : pos+4+int(len)]
+			field_ciphertext = true
+		}
+		pos += 4 + int(len)
+	}
+	if pos != len(b) {
+		return errors.New("encrypted cookie has unexpected length")
+	}
+	if !(field_id && field_nonce && field_ciphertext) {
+		return errors.New("encrypted cookie has missing fields")
+	}
+	return nil
+}
+
+func (c *PlainCookie) Encrypt(key []byte, keyid int) (EncryptedCookie, error) {
+	bits := make([]byte, 16)
+	_, err := rand.Read(bits)
+	if err != nil {
+		return EncryptedCookie{}, err
+	}
+
+	aessiv, err := siv.NewCMAC(key)
+	if err != nil {
+		return EncryptedCookie{}, err
+	}
+
+	b := c.Encode()
+
+	var ecookie EncryptedCookie
+	ecookie.ID = uint16(keyid)
+	ecookie.Nonce = bits
+	ecookie.Ciphertext = aessiv.Seal(nil /* dst */, ecookie.Nonce, b, nil /* additionalData */)
+
+	return ecookie, nil
+}
+
+func (c *EncryptedCookie) Decrypt(key []byte, keyid int) (PlainCookie, error) {
+	if c.ID != uint16(keyid) {
```

Comment:
If you try to decrypt a cookie with the wrong key it would give you this feedback instead of just failing in the actual aead algorithm. Do not know if it is necessary though.

Reply:
I would remove the parameter. The extra-check seems to be more like a potentially useful local wrapper function on the client side. But looking just at this interface, some clients will be tempted to call the method like this: `encryptedCookie.Decrypt(key, encryptedCookie.ID)` which definitely seems a little weird.

Reply:
Done